### ISSUE TITLE
Fails on Node v0.12.0

### ISSUE THREAD

**Issue by silvenon** [assetgraph/assetgraph-builder#162]
Version 3.0.0.

[Demo](https://dl.dropboxusercontent.com/u/22893249/tmp/assetgraph-test.zip).

Command:

```sh
buildProduction --root assetgraph-test --outroot dist
```

```
 ...
 ✔ 0.000 secs: mergeIdenticalAssets
 ✘ ERROR: TypeError: buildProduction transform: minifyAssets transform: Cannot set property length of [object Object] which has only a getter
            at HTMLDocument.core.Node.removeChild (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level1/core.js:662:28)
            at HTMLDocument.<anonymous> (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level2/events.js:370:17)
            at HTMLDocument.proto.(anonymous function) (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/utils.js:23:26)
            at HTMLDocument.inheritFrom.removeChild (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level1/core.js:1667:47)
            at traverse (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Html.js:904:30)
            at Html.extendWithGettersAndSetters.minify (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Html.js:915:10)
            at /usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/transforms/minifyAssets.js:5:23
            at Array.forEach (native)
            at minifyAssets (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/transforms/minifyAssets.js:3:41)
            at Immediate._onImmediate (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/index.js:665:21)
            at processImmediate [as _immediateCallback] (timers.js:358:17)
```

This was the initial error, before I started reducing the test case ([demo](https://dl.dropboxusercontent.com/u/22893249/tmp/assetgraph-test-original.zip)):

```
 ...
 ✔ 0.000 secs: removeRelations
 ✘ ERROR: TypeError: buildProduction transform: externalizeRelations transform: Cannot set property length of [object Object] which has only a getter
            at [object Object].core.Node.removeChild (node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level1/core.js:662:28)
            at [object Object].<anonymous> (node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level2/events.js:370:17)
            at [object Object].proto.(anonymous function) [as removeChild] (node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/utils.js:23:26)
            at HtmlScript.href (node_modules/assetgraph-builder/node_modules/assetgraph/lib/relations/HtmlScript.js:21:23)
            at HtmlScript.Relation.refreshHref (node_modules/assetgraph-builder/node_modules/assetgraph/lib/relations/Relation.js:119:27)
            at JavaScript.<anonymous> (node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Asset.js:500:38)
            at Array.forEach (native)
            at JavaScript.url (node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Asset.js:499:59)
            at node_modules/assetgraph-builder/node_modules/assetgraph/lib/transforms/externalizeRelations.js:7:33
            at Array.forEach (native)
            at externalizeRelations (node_modules/assetgraph-builder/node_modules/assetgraph/lib/transforms/externalizeRelations.js:5:44)
            at Immediate._onImmediate (node_modules/assetgraph-builder/node_modules/assetgraph/lib/index.js:665:21)
            at processImmediate [as _immediateCallback] (timers.js:358:17)
```

This started happening with the exact same setup that previously worked, and it seems to happen with previous versions too, so it may be my fault somewhere and the demo might work for you.

Can you give me some pointers for debugging? I'm desperate :smiley:

**Comment by Munter** [comment]
Oh dear. That doesn't look good.

I've tried reproducing the error with both test cases. Both with a linked assetgraph-builder and the 3.0.0 release. I was unable to reproduce the problem from either of the examples.

@papandreou do you have any intuition as to what might be happening here?

**Comment by silvenon** [comment]
Yeah, I tried everything I could think of, including completely reinstalling all global Node packages, unlinking anything AssetGraph related etc.

Which version of Node are you running? I'm running 0.12.0, maybe that's the problem.

**Comment by silvenon** [comment]
Yep, works on 0.10.36.

**Comment by Munter** [comment]
Yeah. None of us have started looking at getting this into 0.12

**Comment by silvenon** [comment]
Oh, I was unaware how recent it is.

**Comment by ahri** [comment]
I'm getting the same error on an AngularJS app using the Docker instance:

    $ docker run --rm -it  -v "$(pwd)":/app/ -w /app/ assetgraph/assetgraph-builder index.html --outroot /app/foo
    Guessing --root from input files: file:///app/
     ✔ 0.001 secs: logEvents
     ✔ 0.001 secs: registerRequireJsConfig
     ✔ 0.006 secs: registerLabelsAsCustomProtocols
     ✔ 0.647 secs: loadAssets
     ✔ 3.199 secs: populate
     ✘ ERROR: TypeError: buildProduction transform: assumeRequireJsConfigHasBeenFound transform: Cannot set property length of [object Object] which has only a getter
                at HTMLDocument.core.Node.removeChild (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level1/core.js:662:28)
                at HTMLDocument.<anonymous> (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level2/events.js:370:17)
                at HTMLDocument.proto.(anonymous function) (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/utils.js:23:26)
                at HTMLDocument.inheritFrom.removeChild (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level1/core.js:1667:47)
                at Html.parseTree (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Html.js:217:33)
                at Html.extendWithGettersAndSetters.findOutgoingRelationsInParseTree (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Html.js:252:17)
                at Html.outgoingRelations (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Asset.js:572:44)
                at Html.extendWithGettersAndSetters.populate (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Asset.js:631:17)
                at AssetGraph._.extend.addAsset (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/index.js:161:15)
                at JavaScript.<anonymous> (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Asset.js:660:45)
                at Array.forEach (native)
                at JavaScript.extendWithGettersAndSetters.populate (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/assets/Asset.js:631:36)
                at /usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/transforms/assumeRequireJsConfigHasBeenFound.js:9:33
                at Array.forEach (native)
                at assumeRequireJsConfigHasBeenFound (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/transforms/assumeRequireJsConfigHasBeenFound.js:7:80)
                at Immediate._onImmediate (/usr/local/lib/node_modules/assetgraph-builder/node_modules/assetgraph/lib/index.js:665:21)
                at processImmediate [as _immediateCallback] (timers.js:358:17)


Docker info:

    REPOSITORY                      TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
    assetgraph/assetgraph-builder   latest              0ab5be658fa7        3 weeks ago         1.116 GB

If I can be of any assistance providing more detail please let me know! I'm on Ubuntu.

**Comment by TusharaFernando** [comment]
Same issue here - link to travis build:  https://travis-ci.org/Excelian/Angular-Material-ECMA6-Dashboard/jobs/61589899

**Comment by monte-hayward** [comment]
Running v0.12.2, I've run into a series of issues like this. I will try rolling back node version.

Warning: buildProduction transform: externalizeRelations transform: Cannot read property 'replaceChild' of null Use --force to continue.
Warning: buildProduction transform: externalizeRelations transform: Cannot read property 'replaceChild' of null Use --force to continue.
Warning: buildProduction transform: bundleRelations transform: HtmlRelation.attachNodeBeforeOrAfter: Adjacent node has no parentNode. Use --force to continue.
Warning: buildProduction transform: bundleRelations transform: HtmlRelation.attachNodeBeforeOrAfter: Adjacent node has no parentNode. Use --force to continue.
Warning: buildProduction transform: minifyAssets transform: Cannot set property length of [object Object] which has only a getter Use --force to continue.
Warning: buildProduction transform: minifyAssets transform: Cannot set property length of [object Object] which has only a getter Use --force to continue.

**Comment by monte-hayward** [comment]
same issue with v0.11.16
(Like @silvenon , I had also reduced the test case by commenting out parts of buildProduction.js)

$ grunt -dv --stack build
✔ 0.000 secs: removeEmptyStylesheets
Warning: buildProduction transform: inlineRelations transform: Cannot set property length of [object Object] which has only a getter Use --force to continue.
TypeError: buildProduction transform: inlineRelations transform: Cannot set property length of [object Object] which has only a getter
  at [object Object].core.Node.insertBefore (/Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level1/core.js:553:30)
  at [object Object].<anonymous> (/Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level2/events.js:332:20)
  at [object Object].proto.(anonymous function) [as insertBefore] (/Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/utils.js:23:26)
  at [object Object].core.Node.appendChild (/Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/assetgraph/node_modules/jsdom/lib/jsdom/level1/core.js:671:17)
  at HtmlStyle.extendWithGettersAndSetters.inline (/Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/assetgraph/lib/relations/HtmlStyle.js:56:19)
  at /Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/grunt-reduce/node_modules/assetgraph-builder/lib/transforms/buildProduction.js:233:38
  at Array.filter (native)
  at inlineRelations (/Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/grunt-reduce/node_modules/assetgraph-builder/lib/transforms/buildProduction.js:215:20)
  at Immediate._onImmediate (/Users/monte/GoogleDrive/Sites/demo/Fundwindr/node_modules/assetgraph/lib/index.js:668:21)
  at processImmediate [as _immediateCallback] (timers.js:358:17)

**Comment by monte-hayward** [comment]
Results from different versions

0.10.38, v0.10.12, 0.11.5, and 0.11.10,  
dyld: lazy symbol binding failed: Symbol not found: _node_module_register
  Referenced from: /contextify.node
  Expected in: dynamic lookup


0.11.13, 0.11.14
dyld: Symbol not found: __ZN2v86Object3SetENS_6HandleINS_5ValueEEES3_

v0.11.15, 0.11.16, 0.12.0
Warning: buildProduction transform: inlineRelations transform: Cannot set property length of [object Object]

**Comment by papandreou** [comment]
Should be fixed in assetgraph-builder 3.7.0, which also includes support for io.js. Please try it and let me know if it solves the problems you've seen.

**Comment by ahri** [comment]
It works great in docker now!
